fuzz/fuzz_targets/encode_decode.rs

```diff
@@ -2,20 +2,33 @@ extern crate bech32;
 
 use std::convert::TryFrom;
 
+use bech32::Hrp;
+
 fn do_test(data: &[u8]) {
     if data.len() < 1 {
+        println!("early return 0");
         return;
     }
 
-    let hrp_end = (data[0] as usize) + 1;
+    // Get a random valid length hrp 1-83 bytes.
+    let hrp_end = ((data[0] as usize) % 82) + 1;
 
-    if data.len() < hrp_end {
+    // +7 for separator and one six bytes of checksum.
+    if data.len() < hrp_end + 7 {
+        println!("early return 1");
         return;
     }
 
-    let hrp = String::from_utf8_lossy(&data[1..hrp_end])
-        .to_lowercase()
-        .to_string();
+    let mut hrp = [0_u8; 83];
+    for i in 0..hrp_end {
+        // valid ascii range of hrp characters is 33-126
+        hrp[i] = (data[i] % 94) + 33;
+    }
+
+    let hrp = hrp.iter().take(hrp_end).map(|b| unsafe {
+        char::from_u32_unchecked(*b as u32)
+    }).collect::<String>();
```

Comment:
In 599ce6d089904386ac6a34d77a6b3b1fe222393f:

Why is this safe?

Reply:
Oh, I see, you're translating the data in the lines above. I wouldn't do that.

I'd just take the first byte as a "length", try to parse up to that many bytes as an HRP, and use the rest as bech32 data.

Reply:
I'd also assert that the HRP round-trips.

Reply:
> I'd just take the first byte as a "length", try to parse up to that many bytes as an HRP

So that's something I don't understand, even with the tight requirements on the ascii byte range will the fuzzer find data that parses?

Reply:
@tcharding yep. It's kinda magical. Basically it will see that whenever it goes out of bounds, this short-circuits the entire logic chain, so it shouldn't do that (or rather, after doing it once, there is nothing more for it to learn). So it is able to very quickly zero in on valid strings for pretty-much any definition of "valid".

Reply:
Woke up this morning thinking about fuzzing, realized I could throw a panic in the fuzz test to see how long it took to hit (had not yet read your post at that stage), it hits it immediately - that is magic, what genius invented that methodology?